### PR TITLE
Adding error-middleware

### PR DESCRIPTION
**Pull request by lucasnachtigall**
Issue #4

### CODE REVIEW

File: src/api/login/login.ts [hunk 1]
```diff
@@ -28,26 +31,34 @@ router.get('/', verifyToken, (req, res) => {
...
+            if(err) {
+                throw new BadRequestError('Not allowed.');
+            } else {
+                res.json({
+                    message: 'Secure Route',
+                    authData: authData,
+                });
+            }
+        } catch(e) {
+            res.send(e)
```
Comment (gabrielsiedler) [inline src/api/login/login.ts]:
```
// no need for try catch
if (err) {
  throw new BadRequestError('');
}
// no need for else, throw returns

res.json({
  ...
})

```

File: src/api/login/login.ts [hunk 2]
```diff
@@ -28,26 +31,34 @@ router.get('/', verifyToken, (req, res) => {
...
```
Comment (gabrielsiedler) [inline src/api/login/login.ts]: Why are you calling the middleware `verifyToken` AND verifying token on the route?

File: src/api/login/login.ts [hunk 3]
```diff
@@ -74,30 +85,40 @@ router.post('/', checkSchema({
...
-        });
-    } else {
-        res.send('Your credentials are invalid');
+        } else {
+            throw new BadRequestError({
+                error: "Your credentials are invalid"
+              });
+        }
+    } catch(e) {
+        res.send(e)
```
Comment (gabrielsiedler) [inline src/api/login/login.ts]: Never leak error directly from the server.
Return an `Internal error - talk to the admin` in the worst case.

The error-middleware already handles that. ONly throw your error and dont worry about it

File: src/api/login/login.ts [hunk 4]
```diff
@@ -74,30 +85,40 @@ router.post('/', checkSchema({
...
+
+    try {
+        if(typeof authToken !== 'undefined') {
+            req.token= authToken;
+            next();
+        } else {
+            throw new BadRequestError('Not allowed.');
+        }        
+    } catch(e) {
+        res.send(e)
```
Comment (gabrielsiedler) [inline src/api/login/login.ts]: never throw error to the enduser.
no need for try/catch blocks;

The happy path call `next()` and the worst case throw error
here you could use `UnauthorizedError()`;

File: src/server.ts [hunk 5]
```diff
@@ -11,6 +13,9 @@ app.use(bodyParser.urlencoded({
   extended: true
 }));
 
+
+app.use(errorMiddleware);
+
```
Comment (gabrielsiedler) [inline src/server.ts]: add it after the `app.use('/api', masterRouter);`

**Review by gabrielsiedler — CHANGES_REQUESTED** [review]

File: src/api/login/login.ts (hunk 3 above)
Comment (lucasnachtigall) [reply src/api/login/login.ts]: Changes pushed

File: src/api/login/login.ts [hunk 6]
```diff
@@ -8,12 +11,21 @@ class Token {
...
 
 declare global {
     namespace Express {
       interface Request {
-        token: Token
+        token: Token,
+        authData: AuthData
       }
     }
   }
```
Comment (gabrielsiedler) [inline src/api/login/login.ts]: what is that?!?!

File: src/api/login/login.ts [hunk 7]
```diff
@@ -27,78 +39,71 @@ router.get('/', verifyToken, (req, res) => {
...
-        });
-    } else {
-        res.send('Your credentials are invalid');
+        } else {
+            throw new BadRequestError({
+                error: "Your credentials are invalid"
+              });
+        }
+    } catch(e) {
+        res.send(e)
```
Comment (gabrielsiedler) [inline src/api/login/login.ts]: Why try catch? Still leaking the error

**Review by gabrielsiedler — CHANGES_REQUESTED** [review]

**Review by gabrielsiedler — APPROVED** [review]
LGTM

**Pull request merged**